.azure-pipelines/generation-pipeline.yml

```diff
@@ -49,94 +51,227 @@ resources:
+  v1Branch: ${{ parameters.v1BranchPrefix }}/$(Build.BuildId)
+  betaBranch: ${{ parameters.betaBranchPrefix }}/$(Build.BuildId)
 
   # branch scheme above causes the following error for TypeScript because beta branch exists
   # fatal: cannot lock ref 'refs/heads/beta/pipelinebuild/35157': 'refs/heads/beta' exists; cannot create 'refs/heads/beta/pipelinebuild/35157'
-  typeScriptBetaBranch: 'pipelinebuild/beta/$(Build.BuildId)'
-
-steps:
-- pwsh: |
-    git config --global user.email "GraphTooling@service.microsoft.com"
-    git config --global user.name "Microsoft Graph DevX Tooling"
-  displayName: 'Git: set user config'
-
-- checkout: msgraph-metadata
-  displayName: checkout metadata
-  fetchDepth: 1
-  persistCredentials: true
-
-- checkout: self
-  displayName: checkout generator
-  fetchDepth: 1
-  submodules: recursive
-  persistCredentials: true
-
-- checkout: msgraph-sdk-dotnet
-  displayName: checkout dotnet
-  fetchDepth: 1
-  persistCredentials: true
-
-- checkout: msgraph-beta-sdk-dotnet
-  displayName: checkout beta dotnet
-  fetchDepth: 1
-  persistCredentials: true
-
-- checkout: msgraph-sdk-php
-  displayName: checkout php
-  fetchDepth: 1
-  persistCredentials: true
-
-- checkout: msgraph-typescript-typings
-  displayName: checkout typescript
-  fetchDepth: 1
-  persistCredentials: true
-
-- checkout: msgraph-sdk-objc-models
-  displayName: checkout objc
-  fetchDepth: 1
-  persistCredentials: true
-
-- checkout: msgraph-sdk-java
-  displayName: checkout java
-  fetchDepth: 1
-  persistCredentials: true
-
-- checkout: msgraph-beta-sdk-java
-  displayName: checkout beta java
-  fetchDepth: 1
-  persistCredentials: true
-
-- template: common-templates/use-dotnet-sdk.yml
-
-- task: DotNetCoreCLI@2
-  displayName: 'Build Typewriter'
-  inputs:
-    command: 'build'
-    projects: '$(Build.SourcesDirectory)/MSGraph-SDK-Code-Generator/src/Typewriter/Typewriter.csproj'
-    arguments: '--configuration $(buildConfiguration)'
-
-- template: generation-templates/dotnet-v1.yml
-- template: generation-templates/dotnet-beta.yml
-- template: generation-templates/php.yml
-- template: generation-templates/typescript.yml
-- template: generation-templates/objc.yml
-- template: generation-templates/java-v1.yml
-- template: generation-templates/java-beta.yml
\ No newline at end of file
+  typeScriptBetaBranch: ${{ parameters.typeScriptBetaBranchPrefix }}/$(Build.BuildId)
+
+stages:
+- stage: stage_build_and_publish_typewriter
+  dependsOn: [] # remove the implicit dependency to any previous stage
+  jobs:
+  - job: build_and_publish_typewriter
+    steps:
+    - template: generation-templates/build-and-publish-typewriter.yml
+
+- stage: stage_v1_metadata
+  dependsOn: stage_build_and_publish_typewriter
+  condition: succeeded('stage_build_and_publish_typewriter')
+  jobs:
+  - job: v1_metadata
+    steps:
+    - template: generation-templates/checkout-metadata.yml
+    - pwsh: |
+        Write-Host "Success!"
+
+- stage: stage_beta_metadata
```

Comment:
qq for my education, since you only checkout the repo here, are you able to reuse the checkout "state" in subsequence stages? is it because it's keeping the same agent with stage dependencies? or is it for another reason?

Reply:
stage_v1_metadata and stage_beta_metadata stages are pretty much a draft at the moment. Michael will implement the logic in there. The only thing I knew for sure as a step was checking out the metadata repo.

Reply:
oh I get it, and we probably expect the metadata to be added as an artifact so it can be used in subsequent stages. Thanks. I thought there was some kind of black magic happening here.